contracts/CarbonVoteXCore.sol

```diff
@@ -0,0 +1,362 @@
+pragma solidity ^0.4.23;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "./ICarbonVoteXReceiver.sol";
+import "./guard.sol";
+
+
+contract CarbonVoteXCore is DSGuard {
+    // apply SafeMath to uint
+    using SafeMath for uint;
+
+    // events that are emitted after changes.
+    event _Register(
+        address indexed msgSender,
+        bytes32 namespace,
+        bytes32 pollId,
+        uint startBlock,
+        uint endBlock,
+        address tokenAddr
+    );
+    event _WriteAvailableVotes (
+        address indexed msgSender,
+        bytes32 namespace,
+        bytes32 pollId, 
+        address voter, 
+        uint votes
+    );
+    event _SendGas(
+        address indexed msgSender, 
+        bytes32 namespace, 
+        bytes32 pollId,
+        uint value
+    );
+    event _SetReceiver(
+        bytes32 namespace,
+        address receiverAddr
+    );
+    event _CancelReceiver(
+        bytes32 namespace,
+        address receiverAddr
+    );
+
+    struct Poll {
+        // map if a voter has already obtained voting right
+        mapping (address => bool) obtained;
+        // map voter's address to the amount of gas sent
+        mapping (address => uint) gasSentByVoter;
+        uint startBlock;
+        uint endBlock;
+        bytes32 pollId;
+        // the ERC20 token used
+        ERC20 token;
+    }
+
+    // namespace => (pollId => Poll struct)
+    mapping(bytes32 => mapping (bytes32 => Poll)) public polls;
+
+    // namespace => receiver addresses
+    mapping(bytes32 => address) receiverAddr;
+
+    // master's address
+    address private master;
+
+    // map the hashed function name to the signature of the function
+    mapping(bytes32 => bytes4) public restrictedFunctionSig;
+
+    /*
+     * Constructor
+     */
+    // @param master the address, which is able to write votes,
+    // the master address is owned by a backend server
+    constructor (address _master) public {
+        // initiate restrictedFunctionSig mapping;
+        restrictedFunctionSig[keccak256("register")] = this.register.selector;
```

Comment:
as I mentioned in the last review, use enum for the permission type. Otherwise, you have to hard code in multiple places. What if you misspell a permission, but no one found that?  It will cause significant errors.
enum will provide type safety.  Avoid hard-coding if possible.

   enum Permission {Register, SendGas, WriteAvailableVotes}
    mapping(bytes32 => bytes4) public restrictedFunctionSig;
    Permission[] avaiablePermissions;
   
   // also abstracting all components initialization into a function will make it easy to maintain
    function initFunctionSig() {
        avaiablePermissions.push(Permission.Register);
        avaiablePermissions.push(Permission.SendGas);
        functionSig[keccak256(Permission.Register)] = this.register.selector;
        functionSig[keccak256(Permission.SendGas)] = this.sendGas.selector;
    }

for the places (like cancelPermissions( addr)) where you need reference all the permissions,  you can iterate the avaiablePermissions. when we add/delete permissions, you just need change initFunctionSig()  and  enum Permission

Reply:
As we talked,  drop off the enum change, but keep the others

Reply:
added initFunctionSig() and bytes32[] avaiablePermissions;

suggesetion on making an enum Permission {Register, SendGas, WriteAvailableVotes} was rejected due to its limitation in solidity and web3.